subsys/power/device_pm.c

```diff
@@ -0,0 +1,243 @@
+/*
+ * Copyright (c) 2018 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <kernel.h>
+#include <device.h>
+#include <misc/__assert.h>
+
+#define LOG_LEVEL CONFIG_PM_LOG_LEVEL /* From power module Kconfig */
+#include <logging/log.h>
+LOG_MODULE_DECLARE(power);
+
+/* Device PM request type */
+#define DEVICE_PM_ASYNC			(1 << 0)
+
+static void device_pm_callback(struct device *dev,
+			       int retval, void *context, void *arg)
+{
+	u32_t target_state;
+
+	if (retval != 0) {
+		if (*((u32_t *)context) == DEVICE_PM_ACTIVE_STATE) {
+			target_state = DEVICE_PM_SUSPEND_STATE;
+		} else {
+			target_state = DEVICE_PM_ACTIVE_STATE;
+		}
+	} else {
+		target_state = *((u32_t *)context);
+	}
+	k_poll_signal_raise(&dev->config->pm->i_signal, target_state);
+}
+
+static void pm_work_handler(struct k_work *work)
+{
+	struct device_pm *pm = CONTAINER_OF(work,
+					struct device_pm, work);
+	struct device *dev = pm->dev;
+	int ret = 0, result;
+	u8_t target_state;
+	u32_t signaled;
+
+	k_sem_take(&dev->config->pm->lock, K_FOREVER);
+	target_state = dev->config->pm->target_state;
+	switch (dev->config->pm->fsm_state) {
+	case DEVICE_PM_FSM_ACTIVE:
+		if (target_state == DEVICE_PM_SUSPEND_STATE) {
+			dev->config->pm->fsm_state =
+					DEVICE_PM_FSM_STATE_SUSPENDING;
+			/* Set device power state */
+			ret = device_set_power_state(dev, target_state,
+						device_pm_callback, NULL);
+		} else {
+			goto fsm_out;
+		}
+		break;
+	case DEVICE_PM_FSM_STATE_SUSPENDED:
+		if (target_state == DEVICE_PM_ACTIVE_STATE) {
+			dev->config->pm->fsm_state =
+				DEVICE_PM_FSM_STATE_RESUMING;
+			/* Set device power state */
+			ret = device_set_power_state(dev, target_state,
+						device_pm_callback, NULL);
+		} else {
+			goto fsm_out;
+		}
+		break;
+	case DEVICE_PM_FSM_STATE_SUSPENDING:
+	case DEVICE_PM_FSM_STATE_RESUMING:
+		break;
+	default:
+		LOG_ERR("Invalid FSM state!!\n");
+	}
+	k_sem_give(&dev->config->pm->lock);
+
+	if (ret < 0) {
+		LOG_ERR("Set Power state(%d) error(%d)\n",
+					target_state, ret);
+		goto fsm_set_stat;
+	}
+
+	/* Wait for the device callback singnal event */
+	ret = k_poll(&dev->config->pm->i_event, 1, K_FOREVER);
```

Comment:
Here you are blocking the whole workqueue. This is not acceptable. Please look again at the pseudocode I prepared.

Reply:
removed the k_poll from worker.